File: Java/AmazonMedium/LLD/ConsistentHashing.java
import java.util.*;

public class ConsistentHashingLoadBalancer {

    // Represents a server in the system
    private static class Server {
        String name;

        Server(String name) {
            this.name = name;
        }

        @Override
        public String toString() {
            return name;
        }
    }

    // Ring to store server and request mappings
    private final TreeMap<Integer, Server> ring = new TreeMap<>();
    private final int numberOfReplicas; // Number of virtual nodes per server

    public ConsistentHashingLoadBalancer(int numberOfReplicas) {
        this.numberOfReplicas = numberOfReplicas;
    }

    // Add a server to the ring
    public void addServer(Server server) {
        for (int i = 0; i < numberOfReplicas; i++) {
            int hash = getHash(server.name + i); // Create a virtual node
            ring.put(hash, server);
        }
    }

    // Remove a server from the ring
    public void removeServer(Server server) {
        for (int i = 0; i < numberOfReplicas; i++) {
            int hash = getHash(server.name + i);
            ring.remove(hash);
        }
    }

    // Get the server for a given request
    public Server getServer(String request) {
        if (ring.isEmpty()) {
            throw new IllegalStateException("No servers available");
        }
        int hash = getHash(request);
        // Find the closest server in the ring
        Map.Entry<Integer, Server> entry = ring.ceilingEntry(hash);
        if (entry == null) {
            // Wrap around to the first server in the ring
            entry = ring.firstEntry();
        }
        return entry.getValue();
    }

    // Hash function to map keys to the ring
    private int getHash(String key) {
        return key.hashCode(); // Simple hash function for demonstration
    }

    public static void main(String[] args) {
        ConsistentHashingLoadBalancer loadBalancer = new ConsistentHashingLoadBalancer(3);

        // Add servers
        Server server1 = new Server("Server1");
        Server server2 = new Server("Server2");
        Server server3 = new Server("Server3");

        loadBalancer.addServer(server1);
        loadBalancer.addServer(server2);
        loadBalancer.addServer(server3);

        // Simulate requests
        String[] requests = {"Request1", "Request2", "Request3", "Request4", "Request5"};
        for (String request : requests) {
            Server server = loadBalancer.getServer(request);
            System.out.println("Request " + request + " is handled by " + server);
        }

        // Remove a server and see how requests are redistributed
        System.out.println("\nRemoving Server2...");
        loadBalancer.removeServer(server2);

        for (String request : requests) {
            Server server = loadBalancer.getServer(request);
            System.out.println("Request " + request + " is now handled by " + server);
        }
    }
}

File: Python/Consistent_hashing.py
import hashlib
import bisect

class ConsistentHashingLoadBalancer:
    def __init__(self, servers=None, num_replicas=3):
        self.num_replicas = num_replicas
        self.ring = {}
        self.sorted_keys = []
        self.servers = set()
        if servers:
            for server in servers:
                self.add_server(server)

    def _hash(self, key):
        """Return a consistent hash value for the given key."""
        return int(hashlib.md5(key.encode()).hexdigest(), 16)

    def add_server(self, server):
        """Add a server to the hash ring."""
        self.servers.add(server)
        for i in range(self.num_replicas):
            hash_key = self._hash(f"{server}:{i}")
            self.ring[hash_key] = server
            bisect.insort(self.sorted_keys, hash_key)

    def remove_server(self, server):
        """Remove a server from the hash ring."""
        self.servers.discard(server)
        for i in range(self.num_replicas):
            hash_key = self._hash(f"{server}:{i}")
            if hash_key in self.ring:
                del self.ring[hash_key]
                self.sorted_keys.remove(hash_key)

    def get_server(self, key):
        """Get the closest server for the given key."""
        if not self.ring:
            return None
        hash_key = self._hash(key)
        index = bisect.bisect(self.sorted_keys, hash_key) % len(self.sorted_keys)
        return self.ring[self.sorted_keys[index]]

# Example usage
servers = ["192.168.1.1", "192.168.1.2", "192.168.1.3"]
lb = ConsistentHashingLoadBalancer(servers)

print(lb.get_server("request1"))
print(lb.get_server("request2"))

# Adding a new server
lb.add_server("192.168.1.4")
print(lb.get_server("request1"))

# Removing a server
lb.remove_server("192.168.1.2")
print(lb.get_server("request1"))

